### PR TITLE
refactor: local barrier manager use prev epoch to track barrier state

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Split from https://github.com/risingwavelabs/risingwave/pull/14377 as suggested in https://github.com/risingwavelabs/risingwave/pull/14377#issuecomment-1878369468

Refactor includes the following changes:
- use `prev_epoch` as the key of barrier state map. Previously we use `curr_epoch`.
- observe barrier inflight latency when barrier has been collected from all actors. Previously we observe the latency in the collect_barrier rpc handler when we get notified on collecting barriers from all actors.
- maintain failure actor in local barrier worker instead of the managed state.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/task/barrier_manager.rs [hunk 1]
```diff
@@ -292,7 +295,15 @@ impl LocalBarrierWorker {
...
     /// will notice actor's exit while collecting.
     fn notify_failure(&mut self, actor_id: ActorId, err: StreamError) {
-        self.state.notify_failure(actor_id, err)
+        let err = err.into_unexpected_exit(actor_id);
+        if let Some(prev_err) = self.failure_actors.insert(actor_id, err.clone()) {
+            warn!(?actor_id, prev_err = %prev_err.as_report(), "actor error overwritten");
+        }
+        for (fail_epoch, notifier) in self.state.notifiers_await_on_actor(actor_id) {
+            if notifier.send(Err(err.clone())).is_err() {
+                warn!(?fail_epoch, ?actor_id, err = ?err.as_report(), "fail to notify actor failure");
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager.rs:304]: Are we intentionally using `?` for error report here?

**Review by BugenZhao — COMMENTED** [review]
LGTM!

**Review by BugenZhao — APPROVED** [review]

File: src/stream/src/task/barrier_manager.rs [hunk 2]
```diff
@@ -92,34 +92,33 @@ enum LocalBarrierEvent {
...
 
     /// Current barrier collection state.
     state: ManagedBarrierState,
 
     /// Save collect `CompleteReceiver`.
     collect_complete_receiver: HashMap<u64, CompleteReceiver>,
 
-    streaming_metrics: Arc<StreamingMetrics>,
+    /// Record all unexpected exited actors.
+    failure_actors: HashMap<ActorId, StreamError>,
```
Comment (tabVersion) [inline src/stream/src/task/barrier_manager.rs:104]: any reason we wanna keep this?

**Review by tabVersion — APPROVED** [review]
lgtm

**Review by yezizp2012 — APPROVED** [review]
Good job! LGTM!

File: src/stream/src/task/barrier_manager.rs (hunk 2 above)
Comment (wenym1) [reply src/stream/src/task/barrier_manager.rs:104]: Just to be consistent with the original logic. We can save the error and early return the error when handling inject_barrier.

**Pull request merged**